### ISSUE TITLE
MongoError: No more documents in tailed cursor

### ISSUE THREAD

**Issue by petergdoyle** [mafintosh/mongojs#205]
**i am getting two errors**
 i thought it was related to changing up to mongo 3. same thing with mongo 2.6.10 but same thing 
```
db.on('error',function(err) {
   ^
TypeError: Object airshop has no method 'on'
```
**but i am also seeing something worse when streaming on a tailable cursor**

**here is the code** : https://raw.githubusercontent.com/petergdoyle/estreaming/master/node/mongo_connect/streaming_client_mongojs.js

**here is the error**: ```MongoError: No more documents in tailed cursor
at Function.MongoError.create (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/error.js:31:11)
at nextFunction (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/cursor.js:579:50)
at /home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/cursor.js:538:7
at queryCallback (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/cursor.js:215:5)
at Callbacks.emit (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/topologies/server.js:94:3)
at null.messageHandler (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/topologies/server.js:235:23)
at Socket. (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/connection/connection.js:259:22)
at Socket.emit (events.js:95:17)
at Socket. (_stream_readable.js:765:14)
at Socket.emit (events.js:92:17)```

**there is a fixed issues filed at mongo: https://jira.mongodb.org/browse/NODE-435**

**i am using mongojs-1.0.1 now, i was using 0.18.2**

mongojs@1.0.1 node_modules/mongojs
├── thunky@0.1.0
├── each-series@0.2.0
├── to-mongodb-core@2.0.0
├── parse-mongo-url@1.1.0
├── xtend@4.0.0
├── once@1.3.2 (wrappy@1.0.1)
├── pump@1.0.0 (end-of-stream@1.1.0)
├── readable-stream@1.1.13 (isarray@0.0.1, inherits@2.0.1, string_decoder@0.10.31, core-util-is@1.0.1)
└── mongodb-core@1.1.33 (kerberos@0.0.12, bson@0.3.2)

**Comment by shakefu** [comment]
This is due to a change in the [mongodb-core](https://github.com/christkv/mongodb-core) methods. PR #207 resolves this.

**Comment by juuxstar** [comment]
Did you ever get this to work using the latest versions?

**Comment by saintedlama** [comment]
Is the error `MongoError: No more documents in tailed cursor` this still an issue in latest mongojs version? Could you post some example code to reproduce the problem or send a pull request with a test?

**Comment by petergdoyle** [comment]
Sure I will try and get back.

**Comment by petergdoyle** [comment]
I am still getting the issue unless I use "mongojs": "0.18.2" 
```
/home/vagrant/estreaming/node/mongo_connect/streaming_client_mongojs.js:9
db.on('error',function(err) {
   ^
TypeError: Object airshop has no method 'on'
    at Object.<anonymous> (/home/vagrant/estreaming/node/mongo_connect/streaming_client_mongojs.js:9:4)
    at Module._compile (module.js:456:26)
    at Object.Module._extensions..js (module.js:474:10)
    at Module.load (module.js:356:32)
    at Function.Module._load (module.js:312:12)
    at Function.Module.runMain (module.js:497:10)
    at startup (node.js:119:16)
    at node.js:929:3
```

https://github.com/petergdoyle/estreaming/blob/master/node/mongo_connect/package.json

here is the program i am using here: https://github.com/petergdoyle/estreaming/blob/master/node/mongo_connect/streaming_client_mongojs.js 

If i switch to use the latest version of mongojs, here is the dependency list that get installed (i don't need mongoose, that was what i was using before mongojs)
```
JSONStream@1.0.4 node_modules/JSONStream
├── through@2.3.8
└── jsonparse@1.0.0

mongojs@1.0.2 node_modules/mongojs
├── thunky@0.1.0
├── each-series@0.2.0
├── to-mongodb-core@2.0.0
├── parse-mongo-url@1.1.0
├── xtend@4.0.0
├── once@1.3.2 (wrappy@1.0.1)
├── readable-stream@1.1.13 (isarray@0.0.1, inherits@2.0.1, string_decoder@0.10.31, core-util-is@1.0.1)
├── pump@1.0.0 (end-of-stream@1.1.0)
└── mongodb-core@1.2.4 (bson@0.4.7, kerberos@0.0.12)

mongodb@2.0.36 node_modules/mongodb
├── readable-stream@1.0.31 (isarray@0.0.1, inherits@2.0.1, string_decoder@0.10.31, core-util-is@1.0.1)
├── es6-promise@2.1.1
└── mongodb-core@1.2.4 (bson@0.4.7, kerberos@0.0.12)

mongoose@4.0.6 node_modules/mongoose
├── regexp-clone@0.0.1
├── sliced@0.0.5
├── muri@1.0.0
├── hooks-fixed@1.0.1
├── mpromise@0.5.4
├── kareem@1.0.1
├── mpath@0.1.1
├── async@0.9.0
├── ms@0.1.0
├── mquery@1.6.1 (debug@2.2.0, bluebird@2.9.26)
├── mongodb@2.0.34 (readable-stream@1.0.31, mongodb-core@1.2.0)
└── bson@0.3.2 (bson-ext@0.1.10)
```

**Comment by petergdoyle** [comment]
sorry there are two issues. the one i just commented on is already being looked at... but if i take the code out to listen on events 
```

db.on('error',function(err) {
    console.log('database error', err);
});

db.on('ready',function() {
    console.log('database connected');
});
```

I still get the error 

```
/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/cursor.js:251
      throw err;
            ^
MongoError: No more documents in tailed cursor
    at Function.MongoError.create (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/error.js:31:11)
    at /home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/cursor.js:595:54
    at queryCallback (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/wireprotocol/2_6_support.js:100:5)
    at Callbacks.emit (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/topologies/server.js:95:3)
    at null.messageHandler (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/topologies/server.js:246:23)
    at Socket.<anonymous> (/home/vagrant/estreaming/node/mongo_connect/node_modules/mongojs/node_modules/mongodb-core/lib/connection/connection.js:262:22)
    at Socket.emit (events.js:95:17)
    at Socket.<anonymous> (_stream_readable.js:765:14)
    at Socket.emit (events.js:92:17)
    at emitReadable_ (_stream_readable.js:427:10)
```

so the program looks like this now:

```
var mongojs = require('mongojs');
var JSONStream = require('JSONStream');

var url = 'mongodb://localhost:27017/airshop';

var db = mongojs(url, ['results']);

//or ... create a cursor and use the console command
var cursor = db.results.find({}, {}, {tailable:true, timeout:false});

// since all cursors are streams we can just listen for data
cursor.on('data', function(doc) {
    console.log(doc);
});

```

**Comment by saintedlama** [comment]
Thanks for your code demo. We're currently working on a solution for the first missing `on` functionality error. I guess the problem with tailable cursors is somewhere buried in mongodb-core. I'll file a bug report there and get back to you again.

**Comment by saintedlama** [comment]
Sorry for being lacy about the issue. Found out that tailable cursor retry has to be implemented by mongojs and has no support in mongodb-core :hankey:

**Comment by saintedlama** [comment]
Should be fixed in version 2.0 (released just some seconds ago)

**Comment by petergdoyle** [comment]
Thanks!
